src/data.rs

```diff
@@ -80,85 +68,350 @@ impl Token {
+                );
+                msgs.push(format!("upload not completed, we expected {expected} bytes due to the Content-Length header, but received only {}", self.bytes_written));
+                self.errored = true; // we consider this state an error and will prevent renaming
+            }
+        } else {
+            warn!("upload of \"{}\" did not contain Content-Length header, leaving it with partial name", self.filename);
+            msgs.push("your request did not contain the Content-Length header, leaving it with partial name".to_owned());
+            self.errored = true;
+        };
+
+        /* warn about data limit exhaustion */
+        if u64::saturating_sub(self.max_dir_size, self.total_size.load(Ordering::Relaxed)) == 0 {
+            msgs.push("data limit reached while uploading".to_owned());
+        }
+
+        if !self.errored {
+            /* if no errors occured, lets rename the file to its final non-partial name */
+            if let Err(e) = self.dir.mark_upload_final(self.filename).await {
+                error!("Error renaming file after an upload: {e}");
+                msgs.push("error renaming file after a complete upload, you can retry by uploading it again".to_owned());
+            }
+        } else {
+            /* by removing the file from the hash set, we allow its reuploads */
+            let mut lock = self.dir.filenames.lock().await;
+            let name = OsString::from(self.filename);
+            _ = lock.remove(&name);
+        }
+
+        msgs
     }
 }
 
-impl From<Token> for String {
-    fn from(t: Token) -> Self {
-        t.to_string()
+impl<'a> async_std::io::Write for DirectoryFileWriter<'a> {
+    fn poll_write(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> std::task::Poll<std::io::Result<usize>> {
+        /* check if the bytes fit into the size limit */
+        if self.total_size.load(Ordering::Relaxed) > self.max_dir_size {
+            self.get_mut().errored = true;
+            return std::task::Poll::Ready(Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "data limit exceeded",
+            )));
+        }
+
+        /* check if it's not too late */
+        if SystemTime::now() > self.expiration_time {
+            self.get_mut().errored = true;
+            return std::task::Poll::Ready(Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "time limit expired",
+            )));
+        }
+
+        // get a reference for total size
+        let size_ref = self.total_size;
+
+        /* write the bytes */
+        let slf = self.get_mut();
+        let res = async_std::io::Write::poll_write(Pin::new(&mut slf.file), cx, buf);
+
+        /* log the bytes written */
+        if let std::task::Poll::Ready(Ok(size)) = &res {
+            size_ref.fetch_add(*size as u64, Ordering::Relaxed);
```

Comment:
Technically, I think we should first increase the total_size, then write to disk (in case the write failed revert the increase). This way two parallel connections can overshoot the max size by the buffer size (as far as I understand it correctly)

Reply:
:thinking: You are correct, but surprisingly this is not as easy to fix as it sounds... The `poll_write()` function can be called repeatedly for a single write, so naively rearranging the code would account for the bytes multiple times.

I fixed it with a full blown CAS loop "preallocating" the memory block before the actual write. That should hopefully work better.